### PR TITLE
feat: add support for secure TLS bootstrapping to CSE/VHD

### PR DESCRIPTION
**Pull request by cameronmeissner**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
adds support for secure TLS bootstrapping to CSE - when we're ready to fully release this feature I'll create a subsequent PR to add the binary download to the VHD build as well.

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 1]
```diff
@@ -398,8 +397,8 @@ ensureKubelet() {
     mkdir -p "$(dirname "${KUBE_CA_FILE}")"
     echo "${KUBE_CA_CRT}" | base64 -d > "${KUBE_CA_FILE}"
     chmod 0600 "${KUBE_CA_FILE}"
-    
-    if [ "${ENABLE_TLS_BOOTSTRAPPING}" == "true" ]; then
+
+    if [ "${ENABLE_SECURE_TLS_BOOTSTRAPPING}" == "true" ] || [ "${ENABLE_TLS_BOOTSTRAPPING}" == "true" ]; then
```
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/cse_config.sh:401]: git made this hard to read lol, had to see what the current code was doing in vscode xD

File: parts/linux/cloud-init/artifacts/cse_install.sh [hunk 2]
```diff
@@ -71,6 +71,22 @@ downloadCNI() {
...
+    local plugin_url="https://k8sreleases.blob.core.windows.net/aks-tls-bootstrap-client/${plugin_version}/linux/amd64/tls-bootstrap-client"
+    if [[ $(isARM64) == 1 ]]; then
+        plugin_url="https://k8sreleases.blob.core.windows.net/aks-tls-bootstrap-client/${plugin_version}/linux/arm64/tls-bootstrap-client"
+    fi
+
+    mkdir -p /opt/azure/containers
+
+    if [ ! -f "$plugin_download_path" ]; then
+        retrycmd_if_failure 30 5 60 curl -fSL -o "$plugin_download_path" "$kubelet_plugin_url" || exit $ERR_DOWNLOAD_SECURE_TLS_BOOTSTRAP_KUBELET_EXEC_PLUGIN_TIMEOUT
+        chmod 755 "$plugin_download_path"
```
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/cse_install.sh:86]: any reason why we are giving such a broad range of access?

File: parts/linux/cloud-init/artifacts/cse_install.sh [hunk 3]
```diff
@@ -71,6 +71,22 @@ downloadCNI() {
...
 
+downloadSecureTLSBootstrapKubeletExecPlugin() {
+    local plugin_download_path="/opt/azure/containers/tls-bootstrap-client"
+    local plugin_version="v0.1.0-alpha.2"
+    local plugin_url="https://k8sreleases.blob.core.windows.net/aks-tls-bootstrap-client/${plugin_version}/linux/amd64/tls-bootstrap-client"
+    if [[ $(isARM64) == 1 ]]; then
+        plugin_url="https://k8sreleases.blob.core.windows.net/aks-tls-bootstrap-client/${plugin_version}/linux/arm64/tls-bootstrap-client"
+    fi
+
+    mkdir -p /opt/azure/containers
```
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/cse_install.sh:82]: should this not be mkdir -p $plugin_download_path?

Comment (anujmaheshwari1) [reply parts/linux/cloud-init/artifacts/cse_install.sh:82]: wait my bad, thats just a binary right

Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:82]: no you're right, I actually think it should be

mkdir -p $plugin_download_path as you said, with:

local plugin_download_path="/opt/azure/containers"

Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:82]: yeah it's just a binary

File: parts/linux/cloud-init/artifacts/cse_install.sh (hunk 2 above)
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:86]: it really just needs executable access, I can revisit this

File: parts/linux/cloud-init/artifacts/cse_install.sh (hunk 3 above)
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:82]: we could just get the basepath of plugin_download_path and do it that way I guess

File: parts/linux/cloud-init/artifacts/cse_install.sh (hunk 2 above)
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh]: this seems consistent with what we do for directories that contain executables, will keep as is for now I think

File: parts/linux/cloud-init/artifacts/cse_install.sh (hunk 3 above)
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh]: refactored

File: parts/linux/cloud-init/artifacts/cse_cmd.sh [hunk 4]
```diff
@@ -115,6 +115,7 @@ NO_PROXY_URLS="{{GetNoProxy}}"
 PROXY_VARS="{{GetProxyVariables}}"
 ENABLE_TLS_BOOTSTRAPPING="{{EnableTLSBootstrapping}}"
 ENABLE_SECURE_TLS_BOOTSTRAPPING="{{EnableSecureTLSBootstrapping}}"
+CUSTOM_SECURE_TLS_BOOTSTRAP_AAD_SERVER_APP_ID="{{GetCustomSecureTLSBootstrapAADServerAppID}}"
```
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/cse_cmd.sh:118]: keeping this as an optional override for now

**Review by anujmaheshwari1 — APPROVED** [review]

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 5]
```diff
@@ -409,6 +408,41 @@ ensureKubelet() {
 [Service]
 Environment="KUBELET_TLS_BOOTSTRAP_FLAGS=--kubeconfig /var/lib/kubelet/kubeconfig --bootstrap-kubeconfig /var/lib/kubelet/bootstrap-kubeconfig"
 EOF
+    fi
+
+    if [ "${ENABLE_SECURE_TLS_BOOTSTRAPPING}" == "true" ]; then
+        AAD_RESOURCE="6dae42f8-4368-4678-94ff-3960e28e3630"
```
[the remark this answers is not in the document]
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_config.sh:414]: yeeeah it should be fine - I'd rather not to do but the override will essentially give us the ability to correctly set this when we need to on the AKS side

this is also a public value: https://azure.github.io/kubelogin/concepts/aks.html#:~:text=6dae42f8%2D4368%2D4678%2D94ff%2D3960e28e3630

[the remark this answers is not in the document]
Comment (Bryce-Soghigian) [reply parts/linux/cloud-init/artifacts/cse_config.sh:414]: Love a minimal contract :)

**Review by tallaxes — APPROVED** [review]
👍

**Pull request merged**